containers/src/Data/IntMap/Internal.hs

```diff
@@ -3475,69 +3476,127 @@ instance Read1 IntMap where
 -- `linkWithMask` is useful when the `branchMask` has already been computed
-linkWithMask :: Mask -> Prefix -> IntMap a -> IntMap a -> IntMap a
-linkWithMask m p1 t1 {-p2-} t2
-  | zero p1 m = Bin p m t1 t2
-  | otherwise = Bin p m t2 t1
+linkWithMask :: Int -> Key -> IntMap a -> Key -> IntMap a -> IntMap a
+linkWithMask m k1 t1 k2 t2
+  | natFromInt k1 < natFromInt k2 = Bin p t1 t2
+  | otherwise = Bin p t2 t1
   where
-    p = mask p1 m
+    p = Prefix (mask k1 m .|. m)
 {-# INLINE linkWithMask #-}
 
 {--------------------------------------------------------------------
   @bin@ assures that we never have empty trees within a tree.
 --------------------------------------------------------------------}
-bin :: Prefix -> Mask -> IntMap a -> IntMap a -> IntMap a
-bin _ _ l Nil = l
-bin _ _ Nil r = r
-bin p m l r   = Bin p m l r
+
+bin :: Prefix -> IntMap a -> IntMap a -> IntMap a
+bin _ l Nil = l
+bin _ Nil r = r
+bin p l r   = Bin p l r
 {-# INLINE bin #-}
 
 -- binCheckLeft only checks that the left subtree is non-empty
-binCheckLeft :: Prefix -> Mask -> IntMap a -> IntMap a -> IntMap a
-binCheckLeft _ _ Nil r = r
-binCheckLeft p m l r   = Bin p m l r
+binCheckLeft :: Prefix -> IntMap a -> IntMap a -> IntMap a
+binCheckLeft _ Nil r = r
+binCheckLeft p l r   = Bin p l r
 {-# INLINE binCheckLeft #-}
 
 -- binCheckRight only checks that the right subtree is non-empty
-binCheckRight :: Prefix -> Mask -> IntMap a -> IntMap a -> IntMap a
-binCheckRight _ _ l Nil = l
-binCheckRight p m l r   = Bin p m l r
+binCheckRight :: Prefix -> IntMap a -> IntMap a -> IntMap a
+binCheckRight _ l Nil = l
+binCheckRight p l r   = Bin p l r
 {-# INLINE binCheckRight #-}
 
 {--------------------------------------------------------------------
-  Endian independent bit twiddling
+  Branching
 --------------------------------------------------------------------}
 
--- | Should this key follow the left subtree of a 'Bin' with switching
--- bit @m@? N.B., the answer is only valid when @match i p m@ is true.
-zero :: Key -> Mask -> Bool
-zero i m
-  = (natFromInt i) .&. (natFromInt m) == 0
-{-# INLINE zero #-}
+data MapMapBranch
+  = ABL  -- ^ A contains B in the left child
+  | ABR  -- ^ A contains B in the right child
+  | BAL  -- ^ B contains A in the left child
+  | BAR  -- ^ B contains A in the right child
+  | EQL  -- ^ A and B have equal prefixes
+  | NOM  -- ^ A and B have prefixes that do not match
+
+-- | Calculates how two @Bin@s relate to each other by comparing their
+-- @Prefix@es.
+mapMapBranch :: Prefix -> Prefix -> MapMapBranch
+mapMapBranch p1 p2 = case compare pw1 pw2 of
+  LT | pw2 <= (pw1 .|. (pw1-1)) -> ABR
+     | (pw2 .&. (pw2-1)) <= pw1 -> BAL
+     | otherwise                -> NOM
+  GT | pw1 <= (pw2 .|. (pw2-1)) -> BAR
+     | (pw1 .&. (pw1-1)) <= pw2 -> ABL
+     | otherwise                -> NOM
+  EQ                            -> EQL
+  where
+    pw1 = natFromInt (unPrefix p1)
+    pw2 = natFromInt (unPrefix p2)
+{-# INLINE mapMapBranch #-}
```

Comment:
I'm not great at bit fiddling, so I find this rather intimidating. Do you think you could explain how it works? I don't have a good sense of what sorts of bit pattern relationships it's looking for in each test.

Reply:
There are cases to consider here, but the bit fiddling is not too complex.

* `pw .|. (pw-1)` sets every bit below the mask bit to 1. This is the largest key the Bin can have.
* `pw .&. (pw-1)` sets the mask bit and every bit below it to 0. This is the smallest key the Bin can have.

First, we compare the prefixes to each other. Then we compare a prefix against the smallest/largest keys of the other prefix. This gives us enough information to tell how they relate to each other.

* If `pw1` from Bin A is less than `pw2` from Bin B, and `pw2` is <= the greatest key in the Bin A, the right child of Bin A covers the entirety of Bin B.
* ...and so on

Reply:
I keep wondering if there's some way to get away with fewer conditional branches here, and whether doing so would help performance or not. We have only six possible results, but can perform up to four tests. Is there anything more "clever" that would be worth trying? Last I checked, GHC's native code generator really didn't like doing branchless stuff, but maybe there's something else to play with? If so, that obviously doesn't have to be done before the merge.

Reply:
Would you mind putting this explanation in code comments? I think it's brief enough, and a great way to guide future contributions into things.

Reply:
If the concern is that `mapMapBranch` won't inline, then I think the best way forward is to make synonyms for comparisons and unroll the rest of the function at every definition. It's not that much larger than what you have now and the cases are overwhelmingly independent.

```haskell
case compare pA pB of
  EQ                  -> goEQ

  LT | pB <= upper pA -> goABR
     | pA >= lower pB -> goBAL
     | otherwise      -> goANB

  GT | pA <= upper pB -> goBAR
     | pB >= lower pA -> goABL
     | otherwise      -> goBNA
```

There is probably a better way to name the comparisons, though if you know how the tree works they are self-explanatory (`lower`/`upper` are `left`/`right` boundaries respectively and you are querying whether a node is *inside* another node).

Reply:
I don't know if there's anything to optimize in the function itself, it's extremely minimal and it's only downside is that it's completely sequential. I guess you could try to force the evaluation of the "lower" comparison before the "upper" one so they're parallel instruction-wise, but I wouldn't be surprised if that optimization is already performed automatically.

Reply:
Eh, we'll worry about inlining issues later. If there are issues, they're sure to be subtle.

Reply:
> If the concern is that `mapMapBranch` won't inline...

I've marked it `INLINE` and checked the Core to make sure that GHC does its thing and is able to branch on the comparisons directly, so we should be good on that front.